src/main/java/chess/piece/type/Piece.java

```diff
@@ -0,0 +1,84 @@
+package chess.piece.type;
+
+import java.util.Map;
+import java.util.Objects;
+
+import chess.board.Location;
+import chess.score.Score;
+import chess.team.Team;
+
+public abstract class Piece {
+    private static final char BLACk_KING_VALUE = 'K';
+    private static final char WHITE_KING_VALUE = 'k';
+
+    public final char name;
+    protected final Score score;
+
+    Piece(char name, Score score) {
+        this.name = name;
+        this.score = score;
+    }
+
+    public abstract boolean canMove(Location now, Location after);
+
+    public boolean isSameTeam(Team team) {
+        return isBlack() == team.isBlack();
+    }
+
+    public boolean isSameTeam(boolean black) {
+        return isBlack() == black;
+    }
+
+    public boolean isSameTeam(Piece piece) {
+        return isBlack() == piece.isBlack();
+    }
+
+    public boolean isKing() {
+        return this.name == BLACk_KING_VALUE || this.name == WHITE_KING_VALUE;
+    }
+
+    protected boolean isBlack() {
+        return Character.isUpperCase(name);
+    }
+
+    public char getName() {
+        return name;
+    }
+
+    public Score getScore() {
+        return score;
+    }
+
+    public boolean hasObstacle(Map<Location, Piece> board, Location now, Location destination) {
+        Location nowLocation;
+        for (int weight = 1;; weight++) {
+            nowLocation = now.calculateNextLocation(destination, weight);
```

Comment:
```suggestion
            Location nowLocation = now.calculateNextLocation(destination, weight);
```

자료형의 범위를 for문으로 제한한다면 for문 내부만 확인하면 되지만, 범위가 메서드로 늘어난다면 메서드의 코드를 분석해야 하는경우가 발생할 수 있습니다.
또한 범위를 줄임으로써 문제 발생 가능성을 줄일 수 있기 때문에 가능한 좁은 범위를 가지도록 하는 것이 좋습니다.

Reply:
`변수의 범위를 가장 좁은 단위를 가지는 것이 좋다!` 
 - 코드 분석의 부담을 줄이기 위해서
 - 문제 방생 가능성을 줄이기 위해서